src/gtest/test_libsnark.cpp

```diff
@@ -0,0 +1,470 @@
+#include <gtest/gtest.h>
+
+#include "libsnark/algebra/fields/bigint.hpp"
+#include "libsnark/algebra/curves/alt_bn128/alt_bn128_pp.hpp"
+#include "libsnark/reductions/r1cs_to_qap/r1cs_to_qap.hpp"
+#include "libsnark/relations/constraint_satisfaction_problems/r1cs/examples/r1cs_examples.hpp"
+#include "zk_proof_systems/ppzksnark/r1cs_ppzksnark/r1cs_ppzksnark.hpp"
+#include "zk_proof_systems/ppzksnark/r1cs_ppzksnark/examples/run_r1cs_ppzksnark.hpp"
+using namespace libsnark;
+
+TEST(libsnark, bigint)
+{
+    static_assert(ULONG_MAX == 0xFFFFFFFFFFFFFFFFul, "unsigned long not 64-bit");
+    static_assert(GMP_NUMB_BITS == 64, "GMP limb not 64-bit");
+
+    const char *b1_decimal = "76749407";
+    const char *b2_decimal = "435020359732196472065729437602";
+    const char *b3_decimal = "33387554642372758038536799358397002014";
+    const char *b2_binary = "0000000000000000000000000000010101111101101000000110100001011010"
+                            "1101101010001001000001101000101000100110011001110001111110100010";
+
+    bigint<1> b0 = bigint<1>(0ul);
+    bigint<1> b1 = bigint<1>(b1_decimal);
+    bigint<2> b2 = bigint<2>(b2_decimal);
+
+    ASSERT_TRUE(b0.as_ulong() == 0ul);
+    ASSERT_TRUE(b0.is_zero());
+    ASSERT_TRUE(b1.as_ulong() == 76749407ul);
+    ASSERT_TRUE(!(b1.is_zero()));
+    ASSERT_TRUE(b2.as_ulong() == 15747124762497195938ul);
+    ASSERT_TRUE(!(b2.is_zero()));
+    ASSERT_TRUE(b0 != b1);
+    ASSERT_TRUE(!(b0 == b1));
+
+    ASSERT_TRUE(b2.max_bits() == 128);
+    ASSERT_TRUE(b2.num_bits() == 99);
+    for (size_t i = 0; i < 128; i++) {
+        ASSERT_TRUE(b2.test_bit(i) == (b2_binary[127-i] == '1'));
+    }
+
+    bigint<3> b3 = b2 * b1;
+
+    ASSERT_TRUE(b3 == bigint<3>(b3_decimal));
+    ASSERT_TRUE(!(b3.is_zero()));
+
+    bigint<3> b3a { b3 };
+    ASSERT_TRUE(b3a == bigint<3>(b3_decimal));
+    ASSERT_TRUE(b3a == b3);
+    ASSERT_TRUE(!(b3a.is_zero()));
+
+    mpz_t m3;
+    mpz_init(m3);
+    b3.to_mpz(m3);
+    bigint<3> b3b { m3 };
+    ASSERT_TRUE(b3b == b3);
+
+    bigint<2> quotient;
+    bigint<2> remainder;
+    bigint<3>::div_qr(quotient, remainder, b3, b2);
+    ASSERT_TRUE(quotient.num_bits() < GMP_NUMB_BITS);
+    ASSERT_TRUE(quotient.as_ulong() == b1.as_ulong());
+    bigint<1> b1inc = bigint<1>("76749408");
+    bigint<1> b1a = quotient.shorten(b1inc, "test");
+    ASSERT_TRUE(b1a == b1);
+    ASSERT_TRUE(remainder.is_zero());
+    remainder.limit(b2, "test");
+
+    try {
+        (void)(quotient.shorten(b1, "test"));
+        ASSERT_TRUE(false);
+    } catch (std::domain_error) {}
+    try {
+        remainder.limit(remainder, "test");
+        ASSERT_TRUE(false);
+    } catch (std::domain_error) {}
```

Comment:
Also consider using ``EXPECT_*`` instead of ``ASSERT_*`` (to get all failures rather than just the first in a test), and ``EXPECT_EQUAL(a, b)`` instead of ASERT_TRUE(a == b)`` (to print ``a`` and ``b`` in the failure message).